Lua Scripting feature

guyzmo opened the pull request:
Offers a basic API that's working and have some tests.

cf #153 for comments

toogley commented:
Hey.

in general nice work :) I don't use mutt much, but i think it would provide some great readability advantages when we can replace [our conditional features](https://www.neomutt.org/feature/cond-date) with lua. What do you think about that?

guyzmo commented:
please go comment on the main issue, I'd rather keep only code focused discussion here.

toogley commented on main.c:
```diff
@@ -262,9 +265,9 @@ int main (int argc, char **argv, char **environ)
+    if ((i = getopt (argc, argv, "+A:a:Bb:F:f:c:Dd:Ee:g:GH:s:i:hm:npQ:RvxyzZ")) != EOF)
 #else
-    if ((i = getopt (argc, argv, "+A:a:b:F:f:c:Dd:Ee:H:s:i:hm:npQ:RvxyzZ")) != EOF)
+    if ((i = getopt (argc, argv, "+A:a:Bb:F:f:c:Dd:Ee:H:s:i:hm:npQ:RvxyzZ")) != EOF)
```
is that an accidental insertion? It looks like one.

toogley replied:
And in general: What does the string `"+A:a:Bb:F:f:c:Dd:Ee:g:GH:s:i:hm:npQ:RvxyzZ"` do?

flatcap replied:
It the controlling string for `getopt()` which parses the command line.
`B` means there's a `-B` option
`b:` means there's a `-b ARG` option

guyzmo replied:
no it's not accidental, I've added `-B` to add a batch mode. 

But I definitely agree that `getopt` sucks ☺

toogley commented:
* Would it be diffiicult to allow a lua file as only configuration file? Instead of muttrc?
* Do your changes allow also the execution of a lua function in a macro or somewhere else?

somini commented:
I think deprecating the muttrc is a bit too much.
I do agree that automatically sourcing lua files if they are on certain directories would be interesting. Following XDG standards, it would be something like `/usr/lib/mutt/lua:/etc/mutt/lua:$HOME/.config/mutt/lua`

toogley commented:
@somini i didn't mean to deprecate muttrc. I meant that **I** don't have to use it.

However, i agree with the rest of your comment.

toogley commented:
@flatcap why did you close this PR?

toogley commented:
@guyzmo ah, i think that's also because this PR is based on neomutt/neomut (which was deleted)

flatcap commented:
I've just pushed an updated version with a few minor changes:
- Undo the whitespace changes in `init.c`
- Move `test.lua` to `contrib/`
- Run `clang-format` on `mutt_lua.c`

@neomutt/reviewers Please Review

gahr commented:
Could this branch please be rebased, so we can see the actual changes against the current master branch? Thanks!

flatcap commented:
@gahr, I see the problem but I don't know how to fix it.
The branch was created from [neomutt] and the diff is relative to that.
I've updated that branch, but it seems that's not enough.

`git ls-remote` shows me that `refs/pull/298/merge` is probably what needs updating.

guyzmo commented:
@gahr I don't know what to do about that… The first commit for this PR is c27ece34976a2d8b960a1e771684de87bfd6783b, and all the former ones have the same hash as the ones in master. So it looks like it's a github bug, not my fault.

gahr commented:
As discussed on IRC, could you please split `mutt_global_types.h` into two separate header files, one for `option_t` related stuff (say, `option.h`) and one for `command_t` related stuff (say, `command.h`)?

guyzmo commented:
This PR refactors parts of how muttrc gets parsed (by splitting some functions like `parse_set` and exposing `mutt_set_option()` and `mutt_get_option()`). I see that some refactoring work is being done in the area, which is great, but it's dangerously close to things I changed here.

I know @gahr wants to be convinced by the API exposed and wants to see mindblowing scripts as it gets in, but I see nothing wrong in going babysteps and having a small circle of users starting to play around with that API before advertising and documenting it. May I remind that I have things I want to do with a scripting API, that's currently not available because I did not want to expose internal API without thinking about design first.

Most importantly, I personally don't want to keep housekeeping a branch that I'm already having a hard time to keep correctly rebased. I'm seeing #443 as having the potential to be problematic (I'm not saying it is, but sooner or later shit will happen).

So if anybody has comments to do about this patch:

* what can be done to *improve* the API? (issues with namespacing, naming, eventually behaviour)
* is there any issue with the code or the way things are refactored?

gahr commented:
Any impending reasons why you can't PR the refactoring to `mutt_(set|get)_options` separately?

> Most importantly, I personally don't want to keep housekeeping a branch that I'm already having a hard time to keep correctly rebased.

Not a reason for even considering merging, for me.

guyzmo commented:
> Any impending reasons why you can't PR the refactoring to mutt_(set|get)_options separately?

I did my best to squash 30+ commits from the dev branch into a few (7) atomic commits, each of them compile along the way. Though that refactoring took place iteratively through the commit history, and I did not succeed on extracting them as their own commits (like I did with `parse_regex` or adding `-B` which are less scattered across the source). Also those API needed to be extracted *because* I was exposing them with the lua code, and the lua code helped in testing the set/get code to make sure it was working fine.

> > Most importantly, I personally don't want to keep housekeeping a branch that I'm already having a hard time to keep correctly rebased.

> Not a reason for even considering merging, for me.

You're not helping.

I think we're educated enough to judge the expressiveness of an API without needing examples, before writing said examples, so I think the lack of *mindblowing examples* is not a compelling reason for vetoing a merge. That being said, I've started to code a few cool lua stuff with the existing API, but it's very low priority on my TODO list.

A good reason not to merge is faulty, has a bad design, or there are argumented reasons why scripting in lua is wrong. Thus my questions, which I repeat:

* what can be done to improve the API? (issues with namespacing, naming, eventually behaviour)
* is there any issue with the code or the way things are refactored?

gahr commented:
We clearly have to opposite views on this: you are seeking for a good reason *not to merge*; I am seeking for a good reason *to merge*. I said I don't see any, and I'm saying that your "not wanting to keep housekeeping" your WIP branch is not a good reason to merge it.

I don't think it's fair on the other project members to propose a half-baked pile of code, ask for review, complain that it's difficult to keep the feature branch up to date because you have little time or interest in continuing the work, and therefore propose that it's merged as-is.

Bluntly speaking, in my opinion, this branch is not ready for production (hence, master). I think you're senior enough to see it for yourself, but since you're asking, let me give you a small list. It took me several of minutes to prepare this and I really feel it should be on your side to make sure things are tidy.

* is `contrib/test.lua` something you really want to merge?
* I don't see `mutt_commands.h` and `mutt_options.h` appearing in the Makefile
* There is a good amount of code duplication between `mutt_option_set` and `mutt_parse_set`. I understand your only reason for `mutt_option_set` is to be able to call it from Lua code, but sorry this isn't proper refactoring. We now have two unrelated functions that partly do similar things.
* Similarly, the code you extracted into `mutt_command_get` appears in various points in the file. Why not just go one step further and use this new function you added?
* I don't see OP_LUA_RUN being used anywhere (doc/code misalignment)
* autoconf.ac stuff does not work properly. I get `Lua: yes` at the end but mutt_lua.o is not built

gahr commented:
Last but not least - and perhaps more importantly - about examples. 

1. How do you expect users to try to use something where there are zero examples of *what* can be done and *how* to do it?
2. Lack of any examples means that most of the code in `mutt_lua.c` has zero coverage. Are you seriously proposing to merge to master 430 lines of C code that are never executed and ask users to just *play with it*?

guyzmo commented:
> We clearly have to opposite views on this: you are seeking for a good reason not to merge; I am seeking for a good reason to merge. I said I don't see any […]

A good reason to merge is: this adds the building foundation for adding scripting abilities to the code base. I believe we're several who want to see scripting abilities added to neomutt, so here's a good and compelling reason to add it.
Whether there are basic concrete *technical* reasons why not to add this *implementation* of a scripting feature, I'm all ears.

> I don't think it's fair on the other project members to propose a half-baked pile of code […]

I don't think this is half baked, I have tested the feature and it's working. I even wrote a functional test framework based on that API (which looks like it's not a good example to you). And as a senior yourself, you know there's a time where you need other eyes to comment the code, so it can actually be better. You know, feedback…

* is contrib/test.lua something you really want to merge?

Hum… that's a mistake I think has been introduced within a former `git rebase -i`, I meant to publish [code from this gist](https://gist.github.com/guyzmo/9ed8ec8cd54486cf6e97d4c8f2526a2e). Thank you, I'll amend that.

> * I don't see mutt_commands.h and mutt_options.h appearing in the Makefile

ok I'll add that.
As a side question, couldn't we actually use a wildcard glob in the makefile? (feel free to answer on IRC to avoid offtopic discussions here)

> * There is a good amount of code duplication between mutt_option_set and mutt_parse_set. I understand your only reason for mutt_option_set is to be able to call it from Lua code, but sorry this isn't proper refactoring. We now have two unrelated functions that partly do similar things.
> * Similarly, the code you extracted into mutt_command_get appears in various points in the file. Why not just go one step further and use this new function you added?

There is code duplication, but I'd like to complete that part in a separate PR (based on the changes introduced by this PR, rather the other way around). If you (and others) really think I should add the refactor as a new commit within *this* PR, I'm ok to do so.

> * I don't see OP_LUA_RUN being used anywhere (doc/code misalignment)

Hum… that's a leftover from @flatcap's tests. There's no `OP_RUA_RUN` and it shall go away. Taking care of that, thanks.

> * autoconf.ac stuff does not work properly. I get Lua: yes at the end but mutt_lua.o is not built

@flatcap is guilty of the autocrap part 😀

> Lack of any examples means that most of the code in mutt_lua.c has zero coverage. Are you seriously proposing to merge to master 430 lines of C code that are never executed and ask users to just play with it?

err… it's actually a part that *HAS* measured [coverage](https://coveralls.io/github/neomutt/neomutt) (95 lines in `mutt_lua.c` and 546 lines in `init.c`) using this [test](https://github.com/neomutt/neomutt/pull/415/files#diff-0935430c1c315331df916865053648b4) and we could increase the coverage if we're stop arguing about how we're doing tests, and just write tests.

guyzmo commented:
> * How do you expect users to try to use something where there are zero examples of what can be done and how to do it?

I'm exposing the API that's already available in mutt (`get`, `set` variables and `enter-command` commands). What all of that does ***is*** covered by the existing documentation, and what *can* be done with it is up to the imagination of script developers.

gahr commented:
So, 
* I did a quick review of part of your PR (I left `mutt_lua.c` out on purpose)
* You did address some of the issues reported, so the issues were there to begin with
* You failed to address some others
* You blamed somebody else for some others more

If after all this, you're still positive that you want to merge this PR right now as-is just like you were ready to merge before my partial review, well, go ahead. Enough talking from my side already. Let it be noted that I oppose to having this level of quality merged in master.

guyzmo commented:
> * I did a quick review of part of your PR (I left `mutt_lua.c` out on purpose)

what is that purpose?

> * You did address some of the issues reported, so the issues were there to begin with
> * You failed to address some others

what issues did I fail to address?

> * You blamed somebody else for some others more

I'm not blaming, I'm just stating something factual, and said it as a joke. You know that thing at the end of the sentence that's called a smiley? You know, humour, joking, ☺😀😁😊😂🤣

> If after all this, you're still positive that you want to merge this PR right now as-is just like you were ready to merge before my partial review, well, go ahead.

* I want this PR to be merged, yes.
* Do I want low code quality to be merged? No.
* How to make a PR merged? Maybe by having people look at it, give reviews, fix it, rinse, repeat.
* I never merged anything *myself* on master and I don't plan on doing so, I leave that hat to @flatcap.

> Let it be noted that I oppose to having this level of quality merged in master.

I'm sorry but so far you're making remarks about *code quality* whereas you only make *one* point about code. That was about code duplication and I generally agree with you about this, and you're not even considering the ways I'm suggesting to solve this (new commit in the same PR? new PR after the fact?).

> Enough talking from my side already.

I'm not looking for *arguing* about every bits of stuff happening in dev, I just want to move on. I value your input very much when it is elaborate, focused and justified and it's driving me crazy when you're taking a posture.

I think that as ambitious as this feature can be, waiting for 1 month ½ to start reviewing it, knowing that it's been there untouched for most of it isn't great. I took upon myself to keep it rebased on latest code. So either:

1. nobody has comments to make about it, and it's good to merge,
2. code can be improved, and there are comments to be given and code to be fixed,
3. nobody cares about that and I'm dropping the towel.

guyzmo commented:
ok, so far:

fixed all comments @gahr, fixed a few stuff based on @guiniol's feedback. I believe the code is ready for merge, so *again* @neomutt/reviewers please review.

----

About duplication of code introduced here (`mutt_option_set` and `mutt_option_get`): I *chose* not to refactor `parse_set` in this PR, because I wanted to introduce the `lua` scripting feature with the most minimal necessary changes without breaking existing code, and at a second stage refactor `parse_set` to use the new decoupling API, and introducing the the complementing `mutt_option_reset` and `mutt_option_unset`.

But before doing that, I'd like to take advantage of the functional *and* unit testing frameworks so I can start writing tests for validating the changes as I'm starting to move thing saround. Hopefully, for the next release, we won't notice that there's been temporary code deduplication in `init.c`.

I believe merging sooner and often *in between releases* is a safer path than me keeping on coding in a branch, at the risk of impossible merge conflicts. So here's a plan I'm suggesting. If you guys have other ideas I'm all ears.
